v0/destinations/serenytics/transform.js

```diff
@@ -0,0 +1,220 @@
+const { EventType } = require("../../../constants");
+const {
+  CustomError,
+  constructPayload,
+  defaultRequestConfig,
+  defaultPostRequestConfig,
+  getErrorRespEvents,
+  getSuccessRespEvents,
+  getHashFromArrayWithDuplicate,
+  ErrorMessage
+} = require("../../util");
+
+const {
+  CONFIG_CATEGORIES,
+  MAPPING_CONFIG,
+  SERENYTICS_TRACK_EXCLUSION_LIST,
+  SERENYTICS_IDENTIFY_EXCLUSION_LIST,
+  SERENYTICS_PAGE_SCREEN_EXCLUSION_LIST
+} = require("./config");
+const { payloadBuilder } = require("./utils");
+
+const responseBuilder = (STORAGE_URL, payload) => {
+  const response = defaultRequestConfig();
+  response.endpoint = STORAGE_URL;
+  response.method = defaultPostRequestConfig.requestMethod;
+  response.body.JSON = payload;
+  return response;
+};
+
+const trackResponseBuilder = (message, { Config }, payload) => {
+  const STORAGE_URL = Config.storageUrlTrack;
+  const storageUrlEventMapping = getHashFromArrayWithDuplicate(
+    Config.eventToStorageUrlMap,
+    "from",
+    "to",
+    false
+  );
+  const { event } = message;
+  if (!event) {
+    throw new CustomError(
+      `[Serenytics]: event name is required in track call.`,
+      400
+    );
+  }
+  if (!storageUrlEventMapping[event] && !STORAGE_URL) {
+    throw new CustomError(`[Serenytics]: storage url is required.`, 400);
+  }
+  const productList = message.properties?.products;
+  if (productList && Array.isArray(productList)) {
+    const responseList = [];
+    const finalPayload = payload;
+    productList.forEach(product => {
+      const productDetails = constructPayload(
+        product,
+        MAPPING_CONFIG[CONFIG_CATEGORIES.PROPERTY.name]
+      );
+      payload = { ...payload, ...productDetails };
+
+      const storageUrlEventList = storageUrlEventMapping[event];
+      if (storageUrlEventList) {
+        storageUrlEventList.forEach(eventUrl => {
+          const response = responseBuilder(eventUrl, payload);
+          responseList.push(response);
+        });
+      }
+      payload = finalPayload;
+    });
+    return responseList;
```

Comment:
Putting individual products is not correct. Let's say for one `Order Completed`event if there are 3 products we will add three entries with the same `property_order_id` but with different product details. It would be better if we put all the products in one event.

Reply:
Updated